File: vgihan/baekjoon/랜선자르기.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let n, k;
let i = 0;
let cables = [];

rl.on("line", function (line) {
  if (!n) {
    [n, k] = line.split(" ").map((v) => parseInt(v));
    return;
  }
  if (++i < n) {
    cables.push(parseInt(line));
    return;
  }
  cables.push(parseInt(line));
  console.log(solution(cables, k));
  rl.close();
}).on("close", function () {
  process.exit();
});

function solution(cables, k) {
  const maxLength = Math.max(...cables);
  return search(1, maxLength, cables, k);
}

function search(start, end, cables, k) {
  const mid = Math.ceil((start + end) / 2);
  const numOfUnitCable = getNumOfCable(mid, cables);

  if (start === end) return mid;
  if (numOfUnitCable < k) return search(start, mid - 1, cables, k);

  return search(mid, end, cables, k);
}

function getNumOfCable(unitLength, cables) {
  return cables.reduce((num, cableLength) => {
    return num + Math.floor(cableLength / unitLength);
  }, 0);
}

File: thingseong/랜선자르기.py
import sys

K, N = map(int, sys.stdin.readline().split())

lan = []
for _ in range(K):
    lan.append(int(sys.stdin.readline()))


s = 1
e = max(lan)


'''
min이 안되는 이유

2 10
1000
1

1로는 1000개 나옴
'''

while s <= e:
    m = (s + e) // 2
    lines = 0

    for l in lan:
        lines += l // m


    if lines >= N:
        s = m + 1
    else:
        e = m - 1
print(e)
        
